File: clients/client-js/src/scheduling.ts
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
dayjs.extend(utc);

import {
  VexillaFeature,
  VexillaSchedule,
  VexillaScheduleType,
} from "@vexilla/types";
export function isScheduledFeatureActive(feature: VexillaFeature) {
  return isScheduleActive(feature.schedule, feature.scheduleType);
}

export function isScheduleActive(
  schedule: VexillaSchedule,
  scheduleType: VexillaScheduleType
) {
  const now = Date.now();
  return isScheduleActiveWithNow(schedule, scheduleType, now);
}

export function isScheduleActiveWithNow(
  schedule: VexillaSchedule,
  scheduleType: VexillaScheduleType,
  now: number
) {
  if (!scheduleType) {
    return true;
  }

  const currentTime = dayjs.utc(now);

  const startDate = dayjs
    .utc(schedule.start)
    .set("hour", 0)
    .set("minute", 0)
    .set("second", 0)
    .set("millisecond", 0);
  const endDate = dayjs
    .utc(schedule.end)
    .set("hour", 23)
    .set("minute", 59)
    .set("second", 59)
    .set("millisecond", 999);

  if (currentTime.isBefore(startDate) || currentTime.isAfter(endDate)) {
    return false;
  }

  if (schedule.timeType === "none") {
    return true;
  }

  const startTime = dayjs.utc(schedule.startTime);
  const endTime = dayjs.utc(schedule.endTime);

  if (schedule.timeType === "start/end") {
    const startDayWithStartTime = startDate
      .set("hour", startTime.hour())
      .set("minute", startTime.minute())
      .set("second", startTime.second())
      .set("millisecond", startTime.millisecond());

    const endDayWithEndTime = endDate
      .set("hour", endTime.hour())
      .set("minute", endTime.minute())
      .set("second", endTime.second())
      .set("millisecond", endTime.millisecond());

    return (
      currentTime.isAfter(startDayWithStartTime) &&
      currentTime.isBefore(endDayWithEndTime)
    );
  } else if (schedule.timeType === "daily") {
    const currentDayWithStartTime = currentTime
      .set("hour", startTime.hour())
      .set("minute", startTime.minute())
      .set("second", startTime.second())
      .set("millisecond", startTime.millisecond());

    const currentDayWithEndTime = currentTime
      .set("hour", endTime.hour())
      .set("minute", endTime.minute())
      .set("second", endTime.second())
      .set("millisecond", endTime.millisecond());

    return (
      currentTime.isAfter(currentDayWithStartTime) &&
      currentTime.isBefore(currentDayWithEndTime)
    );
  }

  return false;
}

File: clients/swift/Sources/VexillaClient/Scheduling.swift
import Foundation

func isScheduledFeatureActive(feature: Feature) throws -> Bool {
  return try isScheduleActive(schedule: feature.schedule, scheduleType: feature.scheduleType)
}

func isScheduleActive(schedule: Schedule, scheduleType: ScheduleType) throws -> Bool {
  return try isScheduleActiveWithNow(schedule: schedule, scheduleType: scheduleType, now: Date())
}

func isScheduleActiveWithNow(schedule: Schedule, scheduleType: ScheduleType, now: Date) throws -> Bool {
  let nowSeconds = now.timeIntervalSince1970
  var calendar = Calendar(identifier: Calendar.Identifier.iso8601)
  guard let utc = TimeZone(identifier: "UTC") else {
    throw VexillaSchedulingError.couldNotCreateUTCTimezone
  }
  calendar.timeZone = utc

  switch scheduleType {
  case .empty:
    return true

  case .environment, .global:
    let startDate = Date(timeIntervalSince1970: TimeInterval(schedule.start / 1000))
    let startOfStartDate = calendar.startOfDay(for: startDate)

    let endDate = Date(timeIntervalSince1970: TimeInterval(schedule.end / 1000))
    guard let dayAfterEndDate = calendar.date(byAdding: .day, value: 1, to: endDate) else {
      throw VexillaSchedulingError.couldNotAddDayToEndDate
    }
    let endOfEndDate = calendar.startOfDay(for: dayAfterEndDate)

    guard startOfStartDate <= now && endOfEndDate >= now else {
      return false
    }

    switch schedule.timeType {
    case .none:
      return true
    case .startEnd:
      let startOfEndDate = calendar.startOfDay(for: endDate)

      let startDateTimestampWithStartTime = (startOfStartDate.timeIntervalSince1970) + Double(schedule.startTime / 1000)
      let endDateTimestampWithEndTime = (startOfEndDate.timeIntervalSince1970) + Double(schedule.endTime / 1000)

      // print("start: \(startDateTimestampWithStartTime) - now: \(nowSeconds) - end: \(endDateTimestampWithEndTime)")

      return (startDateTimestampWithStartTime ... endDateTimestampWithEndTime).contains(nowSeconds)
    case .daily:

      let todayZeroTime = calendar.startOfDay(for: now).timeIntervalSince1970

      let startTimestamp = todayZeroTime + Double(schedule.startTime / 1000)
      let endTimestamp = todayZeroTime + Double(schedule.endTime / 1000)

      if startTimestamp > endTimestamp {
        return startTimestamp <= nowSeconds || nowSeconds <= endTimestamp
      } else {
        return (startTimestamp ... endTimestamp).contains(nowSeconds)
      }
    }
  }
}
